Tests/DataStreamTests.swift

```diff
@@ -0,0 +1,749 @@
+                }
+            }
+
+        waitForExpectations(timeout: timeout)
+
+        // Then
+        XCTAssertTrue(firstStreamOnMain)
+        XCTAssertTrue(firstCompleteOnMain)
+        XCTAssertEqual(firstResponse?.statusCode, 200)
+        XCTAssertTrue(!firstAccumulatedData.isEmpty)
+        XCTAssertTrue(secondStreamOnMain)
+        XCTAssertTrue(secondCompleteOnMain)
+        XCTAssertEqual(secondResponse?.statusCode, 200)
+        XCTAssertNotNil(decodedResponse)
+        XCTAssertNil(decodingError)
+    }
+
+    func testThatDataStreamRequestProducesWorkingInputStream() {
+        // Given
+        let expect = expectation(description: "stream complete")
+
+        // When
+        let stream = AF.streamRequest(URLRequest.makeHTTPBinRequest(path: "xml", headers: [.contentType("application/xml")]))
+            .responseStream { stream in
+                switch stream.event {
+                case .complete:
+                    expect.fulfill()
+                default: break
+                }
+            }.asInputStream()
+
+        waitForExpectations(timeout: timeout)
+
+        // Then
+        let parser = XMLParser(stream: stream)
+        let parsed = parser.parse()
+        XCTAssertTrue(parsed)
+        XCTAssertNil(parser.parserError)
+    }
+
+    func testThatDataStreamCanBeManuallyResumed() {
+        // Given
+        let session = Session(startRequestsImmediately: false)
+        var response: HTTPURLResponse?
+        var streamOnMain = false
+        var completeOnMain = false
+        let expect = expectation(description: "stream complete")
+
+        // When
+        session.streamRequest(URLRequest.makeHTTPBinRequest(path: "stream/1"))
+            .responseStream { stream in
+                switch stream.event {
+                case .stream:
+                    streamOnMain = Thread.isMainThread
+                case let .complete(completion):
+                    completeOnMain = Thread.isMainThread
+                    response = completion.response
+                    expect.fulfill()
+                }
+            }.resume()
+
+        waitForExpectations(timeout: timeout)
+
+        // Then
+        XCTAssertTrue(streamOnMain)
+        XCTAssertTrue(completeOnMain)
+        XCTAssertEqual(response?.statusCode, 200)
+    }
+
+    func testThatDataStreamIsAutomaticallyCanceledOnStreamErrorWhenEnabled() {
+        var response: HTTPURLResponse?
+        var complete: DataStreamRequest.Completion?
+        let expect = expectation(description: "stream complete")
+
+        // When
+        AF.streamRequest(URLRequest.makeHTTPBinRequest(path: "bytes/50"), automaticallyCancelOnStreamError: true)
+            .responseStreamDecodable(of: HTTPBinResponse.self) { stream in
+                switch stream.event {
+                case let .complete(completion):
+                    complete = completion
+                    response = completion.response
+                    expect.fulfill()
```

Comment:
What exactly trips the error in this test? Not obvious where it comes from. Just curious.

Reply:
Returning 50 random bytes won't properly parse as an `HTTPBinResponse`, so the serializer produces an error.

Reply:
Ah thanks. Was thinking too hard on it and missed the obvious.